File: Data Structures & Algorithms/regular-expression-matching/submission-7.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        N,M = len(s),len(p)
        dp = {}
        def dfs(i,j):
            if j == M:
                return i == N
            if (i,j) in dp:
                return dp[(i,j)]
            isMatch = i<N and (s[i] == p[j] or p[j] == ".")
            if j+1<M and p[j+1] == "*":
                dp[(i,j)] = dfs(i,j+2) or (isMatch and dfs(i+1,j))
            elif isMatch:
                dp[(i,j)] = dfs(i+1,j+1)
            else:
                dp[(i,j)] = False
            return dp[(i,j)]
        return dfs(0,0)



        

File: Data Structures & Algorithms/reconstruct-flight-path/submission-7.java
class Solution {
    Map<String,List<String>> adj = new HashMap<>();
    List<String> res = new ArrayList<>();
    public List<String> findItinerary(List<List<String>> tickets) {
        tickets.sort((a,b) -> a.get(1).compareTo(b.get(1)));
        for(List<String>ticket:tickets){
            String src = ticket.get(0),dest = ticket.get(1);
            adj.putIfAbsent(src,new ArrayList<>());
            adj.get(src).add(dest);
        }
        dfs("JFK");
        Collections.reverse(res);
        return res;
    }

    public void dfs(String src){
        if(!adj.containsKey(src)){
            res.add(src);
            return;
        }
        while(!adj.get(src).isEmpty()){
            String dest = adj.get(src).remove(0);
            dfs(dest);
        }
        res.add(src);
    }




        // JFK <-> SEA
        // -> Hou

        // JFk <->
        
        
        // JFK - [Hou,Sea]
        // SEA - [JFK]
        // Hou - []

        // JFK - [Hou,SEA] [SEA]
        // Hou - []
        // JFK - [SEA] []
        // SEA - [JFK] []
        // JFK - []




        // HOU,JFK,SEA,JFK - reverse

        // start with JFK go through each nei in adj for the node and remove from adj 
        //while consuming each nei recursively and backtrack and then add
        // to res


        
        




}
